### PR TITLE
Feature/visibility changed annotation

### PR DESCRIPTION
**Pull request by eboudrant**
Visibility events in Epoxy Inspired from Litho : https://fblitho.com/docs/visibility-handling

### CODE REVIEW

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyModel.java [hunk 1]
```diff
@@ -155,6 +155,25 @@ public void bind(@NonNull T view, @NonNull EpoxyModel<?> previouslyBoundModel) {
...
+  public void visibilityFocusedVisible(@NonNull T view) {
+  }
+
+  public void visibilityInvisible(@NonNull T view) {
+  }
+
+  public void visibilityUnfocusedVisible(@NonNull T view) {
+  }
+
+  public void visibilityFullImpressionVisible(@NonNull T view) {
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyModel.java]: do you have a definition for `FullImpressionVisible`

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyRecyclerView.java [hunk 2]
```diff
@@ -175,6 +191,7 @@ public EpoxyRecyclerView(Context context, @Nullable AttributeSet attrs, int defS
   protected void init() {
     setClipToPadding(false);
     initViewPool();
+    setVisibilityTrackingEnabled(true);
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyRecyclerView.java]: default true ?

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+   *
+   * @param view the current {@link com.airbnb.epoxy.EpoxyViewHolder}'s itemView
+   * @param parent the {@link android.support.v7.widget.RecyclerView}
+   * @param orientation the {@link android.support.v7.widget.LinearLayoutManager}'s orientation
+   *
+   * @return true if the view has been measured
+   */
+  boolean update(@NonNull View view, @NonNull RecyclerView parent, int orientation) {
+    view.getLocalVisibleRect(localVisibleRect);
+    vertical = orientation == LinearLayoutManager.VERTICAL;
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java]: Do you support only LinearLayoutManager ?

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+ * <p>
+ * - VisibilityChangedEvent: this event is triggered when the visibility of the Component on the
+ * screen changes.
+ */
+class EpoxyVisibilityTracker {
+
+  /** Maintain visibility item indexed by view id (identity hashcode) */
+  private final SparseArray<EpoxyVisibilityItem> visibilityIdToItemMap = new SparseArray<>();
+
+  /** listener use to process scroll, layout and attaches events */
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: typo : `used`

Comment (frannois) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: typo : `attach`

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  private final Listener listener = new Listener();
+
+  @Nullable
+  private RecyclerView attachedRecyclerView = null;
+
+  @Nullable
+  private LinearLayoutManager attachedLinearLayoutManager = null;
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: I think this answered my previous question : so no support for non-LinearLayoutManager. Do you have code to no enable it in that case

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    recyclerView.removeOnChildAttachStateChangeListener(this.listener);
+    attachedRecyclerView = null;
+    attachedLinearLayoutManager = null;
+  }
+
+  @Nullable
+  private LinearLayoutManager getLinearLayoutManager() {
+    if (attachedLinearLayoutManager == null && attachedRecyclerView != null) {
+      LayoutManager lm = attachedRecyclerView.getLayoutManager();
+      if (lm instanceof LinearLayoutManager) {
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: maybe a log to say this doesn't support anything other than LinearLayoutManager

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        attachedLinearLayoutManager = (LinearLayoutManager) lm;
+      }
+    }
+    return attachedLinearLayoutManager;
+  }
+
+  private void processRecyclerView(@NonNull RecyclerView recyclerView) {
+    LinearLayoutManager llm = getLinearLayoutManager();
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: nit : final

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+      for (int i = 0; i < llm.getChildCount(); i++) {
+        View child = llm.getChildAt(i);
+        if (child != null) {
+          processChild(child);
+        }
+      }
+    }
+  }
+
+  private void processChild(@NonNull View child) {
```
Comment (frannois) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: do you need this method ?

File: epoxy-processor/src/main/java/com/airbnb/epoxy/GeneratedModelWriter.kt [hunk 9]
```diff
@@ -435,6 +444,18 @@ internal class GeneratedModelWriter(
         }
     }
 
+    private fun generateVisibilityMethods(modelInfo: GeneratedModelInfo): Iterable<MethodSpec> {
+        val methods = ArrayList<MethodSpec>()
+        val boundObjectParam = ParameterSpec.builder(modelInfo.modelType, "object", FINAL).build()
```
Comment (frannois) [inline epoxy-processor/src/main/java/com/airbnb/epoxy/GeneratedModelWriter.kt]: any overhead on adding all of this ? Can it be optional ?

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyRecyclerView.java (hunk 2 above)
Comment (eboudrant) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyRecyclerView.java]: no, it was temporary for testing

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java (hunk 3 above)
Comment (eboudrant) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityItem.java]: yes. btw a `GridLayoutManager` is also a `LinearLayoutManager `

File: epoxy-processor/src/main/java/com/airbnb/epoxy/GeneratedModelWriter.kt (hunk 9 above)
Comment (eboudrant) [reply epoxy-processor/src/main/java/com/airbnb/epoxy/GeneratedModelWriter.kt]: This code get executed at build time, no overhead at runtime. Is it answering your comment ?

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java (hunk 8 above)
Comment (eboudrant) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: yes, from `onChildViewAttachedToWindow` / `onChildViewDetachedFromWindow`

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java (hunk 6 above)
Comment (jess-anders) [reply epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.java]: I would crash if they enable it and it's not a LinearLayoutManager with an IllegalStateException and helpful error message.

File: epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyModel.java [hunk 10]
```diff
@@ -155,6 +155,25 @@ public void bind(@NonNull T view, @NonNull EpoxyModel<?> previouslyBoundModel) {
   public void unbind(@NonNull T view) {
```
Comment (jess-anders) [inline epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyModel.java]: looks nice!  Would make sure performance is good.

**Pull request closed without merging**